services/gitlab/gitlab-merge-requests.service.js

```diff
@@ -0,0 +1,359 @@
+        project: 'gitlab-org/gitlab',
+      },
+      queryParams: {
+        labels: 'test,type::feature',
+        gitlab_url: 'https://gitlab.com',
+      },
+      staticPreview: {
+        label: 'test,failure::new merge requests',
+        message: '68 merged',
+        color: 'brightgreen',
+      },
+      documentation: documentation + labelDocumentation,
+    },
+  ]
+
+  static defaultBadgeData = { label: 'merge requests' }
+
+  static render({ variant, raw, labels, mergeRequestCount }) {
+    const state = variant
+    const isMultiLabel = labels && labels.includes(',')
+    const labelText = labels ? `${isMultiLabel ? `${labels}` : labels} ` : ''
+
+    let labelPrefix = ''
+    let messageSuffix = ''
+    if (raw !== undefined) {
+      labelPrefix = `${state} `
+    } else {
+      messageSuffix = state
+    }
+    const message = `${mergeRequestCount > 10000 ? 'more than ' : ''}${metric(
+      mergeRequestCount
+    )}${messageSuffix ? ' ' : ''}${messageSuffix}`
+    return {
+      label: `${labelPrefix}${labelText}merge requests`,
+      message,
+      color: mergeRequestCount > 0 ? 'brightgreen' : 'yellow',
+    }
+  }
+
+  async fetch({ project, baseUrl, variant, labels }) {
+    let state = variant
+    if (state === 'open') {
+      state = 'opened'
+    }
+    const labelsParam = labels
+      ? {
+          labels,
+        }
+      : {}
+    // https://docs.gitlab.com/ee/api/merge_requests.html#list-project-merge-requests
+    const { res } = await this._request(
+      this.authHelper.withBearerAuthHeader({
+        url: `${baseUrl}/api/v4/projects/${encodeURIComponent(
+          project
+        )}/merge_requests`,
+        options: {
+          searchParams: {
+            state,
+            page: '1',
+            per_page: '1',
+            ...labelsParam,
+          },
+        },
+        errorMessages: {
+          404: 'project not found',
+        },
+      })
+    )
+    const data = this.constructor._validate(res.headers, schema)
+    let mergeRequestCount
+    if (data['x-total'] !== undefined) {
+      mergeRequestCount = data['x-total']
+    } else {
+      // https://docs.gitlab.com/ee/api/index.html#pagination-response-headers
+      // For performance reasons, if a query returns more than 10,000 records, GitLab doesn’t return `x-total` header.
+      // Displayed on the page as "more than 10k".
+      mergeRequestCount = 10001
+    }
+
+    // The total number of MR is in the `x-total` field in the headers.
+    // https://docs.gitlab.com/ee/api/index.html#other-pagination-headers
+    return { mergeRequestCount }
+  }
+
+  async handle(
+    { variant, raw, project },
+    { gitlab_url: baseUrl = 'https://gitlab.com', labels }
+  ) {
+    const { mergeRequestCount } = await this.fetch({
+      project,
+      baseUrl,
+      variant,
+      labels,
+    })
+    return this.constructor.render({
+      variant,
+      raw,
+      labels,
+      mergeRequestCount,
+    })
+  }
+}
```

Comment:
Somewhat conversely to the `render` function, I feel like the `fetch` function is unnecessarily verbose and could be cleaned up a bit. Also note for future reference that there's no need to check for truthiness for url query params, you can just toss the value along and the http handling code farther up the pipe will handle that automatically.

I started out with a suggested change to just `fetch` but ended up wanting to suggest a broader refactoring that I think reads a bit cleaner. It may seem like a large diff, but it's basically just small tweaks/clean up on the `fetch` function, pulling out a separate `render` function, and not wrapping results in an object and forcing destructuring at the call sites.

```suggestion
  static transform(data) {
    if (data['x-total'] !== undefined) {
      return data['x-total']
    } else {
      // https://docs.gitlab.com/ee/api/index.html#pagination-response-headers
      // For performance reasons, if a query returns more than 10,000 records, GitLab doesn’t return `x-total` header.
      // Displayed on the page as "more than 10k".
      return 10001
    }
  }

  async fetch({ project, baseUrl, variant, labels }) {
    // https://docs.gitlab.com/ee/api/merge_requests.html#list-project-merge-requests
    const { res } = await this._request(
      this.authHelper.withBearerAuthHeader({
        url: `${baseUrl}/api/v4/projects/${encodeURIComponent(
          project
        )}/merge_requests`,
        options: {
          searchParams: {
            state: variant == 'open' ? 'opened' : variant,
            page: '1',
            per_page: '1',
            labels,
          },
        },
        errorMessages: {
          404: 'project not found',
        },
      })
    )
    return this.constructor._validate(res.headers, schema)
  }

  async handle(
    { variant, raw, project },
    { gitlab_url: baseUrl = 'https://gitlab.com', labels }
  ) {
    const data = await this.fetch({
      project,
      baseUrl,
      variant,
      labels,
    })
    const mergeRequestCount = this.constructor.transform(data)
    return this.constructor.render({
      variant,
      raw,
      labels,
      mergeRequestCount,
    })
  }
}
```